Finish tinywl.

jsoo1 opened the pull request:
Hello,

@LambdaCalrissian and me have been working on finishing the tinywl implementation. We have something close to a working version here with a few bugs to fix.

We completed the rest of the `failwith`s in the file and completed the ffi required to accomplish it.

There are certainly things here that might need discussion and, as mentioned it still has bugs.

Thanks for the great start!

LambdaCalrissian committed:
Adding role to xdg_shell_surface

LambdaCalrissian committed:
Adding surface events

LambdaCalrissian committed:
Work in progress filling out TinyWL

jsoo1 committed:
wlroots: Refer to implicit t in xdg_shell.surface sig.

jsoo1 committed:
Remove CTypes from tinywl.

* Get back pure ocaml type from Xdg_shell.Surface.role.

jsoo1 committed:
Add keyboard_state field to Seat.

jsoo1 committed:
Rename Xdg_shell.Surface -> Xdg_shell.Xdg_surface.

jsoo1 committed:
Add surface field to Xdg_surface.

jsoo1 committed:
Seat: Return option for focused_surface.

jsoo1 committed:
xdg_surface: Add from_surface.

* Corresponds to wlr_xdg_surface_from_surface in xdg_shell.

jsoo1 committed:
xdg_surface: Move into separate directory.

jsoo1 committed:
wlr_xdg_toplevel: Add wlr_xdg_toplevel_set_activated.

jsoo1 committed:
[WIP] tinywl: Set previous surface to deactivated on xdg_surface.mapped.

LambdaCalrissian committed:
Fixed type of modifiers field

LambdaCalrissian committed:
Expose the Keyboard_modifiers module.

LambdaCalrissian committed:
Add the Keycodes type and module.

Keycodes is not in Wlroots, but CTypes does not support abstract arrays. So
we're using this as a wrapper with a void*.

LambdaCalrissian committed:
Format Wlroots modules on new lines for better commits.

LambdaCalrissian committed:
Add Keycodes, Num_keycodes, and Modifier fields to Keyboard.

LambdaCalrissian committed:
tinywl: Finish focus_view.

jsoo1 committed:
Add keyboard, modifiers struct and keyboard, modifiers fields.

Needed to do keyboard_notify_enter.

jsoo1 committed:
Add keyboard_notify_enter.

This is to finish focus_view in tinywl.

jsoo1 committed:
[WIP] tinywl: Notify keyboard enter in focus_view.

LambdaCalrissian committed:
Add mapped property to view.

LambdaCalrissian committed:
Make view a ref.

We need to mutate this value after it's held, so we need to make it into a ref.

LambdaCalrissian committed:
Add unmap callback.

LambdaCalrissian committed:
xdg_toplevel: Creates types, interfaces, and events

Needed for move and resize events in tinywl.

LambdaCalrissian committed:
view:mutability Change from a ref to mutable on mapped.

Don't need to make the entire thing a ref. Just make the specific field we are
concerned with mutable.

LambdaCalrissian committed:
tinywl: Stub out begin interactive.

Need to handle the interactive events (moving or resizing a winodw).

Need a new data type for the what the cursor is trying to do.

LambdaCalrissian committed:
tinywl:new_xdg_surface Comment about making mapped true.

Might never be set to true. Do we need to handle this?

jsoo1 committed:
tinywl: Begin implementation of server_new_keyboard.

LambdaCalrissian committed:
tinywl:server_new_sdg_surface Remove unneeded reference comment.

jsoo1 committed:
tinywl: Set repeat info in server_new_keyboard.

jsoo1 committed:
keyboard: Add modifiers event, create Events module.

jsoo1 committed:
tinywl: Stub keyboard handlers.

jsoo1 committed:
seat: Add set_keyboard and seat_keyboard_notify_modifiers.

jsoo1 committed:
tinwyl: Add keyboard_handle_modifiers.

jsoo1 committed:
keyboard: Add get_modifiers.

jsoo1 committed:
types: Add wlr_keyboard_modifier.

jsoo1 committed:
keyboard_modifiers: Add has_alt.

jsoo1 committed:
[WIP] tinywl: Stub handle_keybinding

jsoo1 committed:
tinywl: Only do new_xdg_surface handling if TopLevel.

jsoo1 committed:
seat: Add keyboard_notify_key.

jsoo1 committed:
Fill in handle_keybinding

jsoo1 committed:
keyboard_modifiers: Fix has_alt coercions.

LambdaCalrissian committed:
cursor_move: Added ffi and functions.

Amended after hunting down some module shenanigans. Need to have
Event_pointer_motion below Input_device or else you get some real weird
module reference, makes you feel a bit crazy.

Also moved Event_pointer_motion to it's own module. Needed to get the
input device from the event. Had cyclical dependency if the event was on
the pointer module.

LambdaCalrissian committed:
cursor: Add x & y fields

LambdaCalrissian committed:
[WIP] tinywl: process_cursor_motion

LambdaCalrissian committed:
edges: Add module

LambdaCalrissian committed:
tinywl: process_cursor_resize Added edge processing.

* Move discard to toplevel.

LambdaCalrissian committed:
ffi: Add wlr_xdg_surface_surface_at

LambdaCalrissian committed:
ffi: Add wlr_xcursor_manager_set_cursor_image

LambdaCalrissian committed:
tinywl: Finish server_cursor_motion

jsoo1 committed:
ffi: Add wlr_xdg_toplevel_set_size.

jsoo1 committed:
ffi: Add wlr_xdg_surface_get_geometry.

jsoo1 committed:
tinywl: Change Tinywl.box to Box.t.

LambdaCalrissian committed:
ffi:seat Add wlr_seat_notify_pointer_enter

LambdaCalrissian committed:
ffi:seat Add wlr_seat_pointer_clear_focus

LambdaCalrissian committed:
types:pointer_state Add focused_surface

LambdaCalrissian committed:
ffi:seat Add wlr_seat_pointer_notify_motion

jsoo1 committed:
ffi: Add wlr_cursor_warp_absolute.

jsoo1 committed:
types: Add device field to Event_pointer_motion_absolute.

jsoo1 committed:
types: Add x and y fields to wlr_event_pointer_motion_absolute.

jsoo1 committed:
lib/event_pointer_motion_absolute: Move module out of Pointer.

This matches the module Event_pointer_motion.

jsoo1 committed:
event_pointer_motion_absolute: Add time_msec field.

jsoo1 committed:
tinywl: Finish server_cursor_motion_absolute.

jsoo1 committed:
wlroots: Move Event_pointer_button to separate module.

This matches event_pointer_motion*.

jsoo1 committed:
event_pointer_button: Add time_msec field.

jsoo1 committed:
event_pointer_button: Add button field.

jsoo1 committed:
types: Add Button_state.

jsoo1 committed:
event_pointer_button: Add state field.

jsoo1 committed:
ffi: Add wlr_seat_pointer_notify_button.

jsoo1 committed:
tinywl: Finish server_cursor_button.

* Change type of desktop_view_at.
* Make st.cursor_mode mutable.

jsoo1 committed:
types: Add wlr_axis_source enum.

jsoo1 committed:
types: Add wlr_axis_orientation enum.

jsoo1 committed:
event_pointer_axis: Move Event_pointer_axis module to separate file.

jsoo1 committed:
types: Add field orientation of wlr_event_pointer_axis.

jsoo1 committed:
types: Add delta field of wlr_event_pointer_axis.

jsoo1 committed:
types: Add delta_discrete field of wlr_event_pointer_axis.

jsoo1 committed:
types: Add source field of wlr_event_pointer_axis.

jsoo1 committed:
ffi: Add wlr_seat_pointer_notify_axis.

jsoo1 committed:
types: Add field time_msec of wlr_event_pointer_axis.

jsoo1 committed:
tinywl: Finish server_cursor_axis.

jsoo1 committed:
ffi: Add wlr_seat_pointer_notify_frame.

jsoo1 committed:
tinywl: Finish server_cursor_frame.

jsoo1 committed:
tiny: Make view.mapped = true when xdg_surface mapped event happens.

* Remove listener from focus_event.

jsoo1 committed:
tinywl: Create more descriptive failwiths.

* Change print_endline back to failwith

jsoo1 committed:
ffi: Add wlr_output_effective_resolution.

jsoo1 committed:
types: Add wlr_surface_iterator_func_t.

jsoo1 committed:
ffi: Add wlr_xdg_surface_for_each_surface.

jsoo1 committed:
tinywl: Finish output_frame.

* Adds render_frame as subroutine of output_frame

jsoo1 committed:
ffi: Add wlr_surface_get_texture.

jsoo1 committed:
ffi: Add wlr_output_layout_output_coords.

jsoo1 committed:
types: Add wlr_output.scale.

jsoo1 committed:
types: Define enum for wl_output_transform.

jsoo1 committed:
ffi: Add wlr_output_transform_invert.

jsoo1 committed:
ffi: Add wlr_renderer_render_texture_with_matrix.

jsoo1 committed:
ffi: Add wlr_output_render_software_cursors.

wlr_output_render_software_cursors requires libpixman which has no
ocaml bindings yet.  However this is ok for now, since only tinywl.ml
is using it.  Some minimal attempts at adding bindings for pixman
should be taken soon to make this ok.

jsoo1 committed:
tinywl: Finish begin_interactive.

* Moves option elimination outside of process_cursor* functions.
* Changes grab to have an optional resize field, since resizing and
  moving are two disjoint states.
* Changes Resize cursor_mode paramter to Edges.t

jsoo1 committed:
tinywl: Swap out discard for ignore.

jsoo1 committed:
tinywl: Use Module.(expr) to cleanup module noise.

jsoo1 committed:
tinywl: Use single branched i.t.e. where applicable.

jsoo1 committed:
types: Swaps Edges.t from enum to bitwise_enum32

jsoo1 committed:
utils: Add bitwise_enum32.

Edges are used in 32bit fashion. This could be changed since I do not
know if they can be coerced safely to 64 bits.

jsoo1 committed:
types: Add xdg_toplevel_move_event and xdg_toplevel_resize_event.

Armael commented:
Wow, thanks a lot for this work, that looks great! I appreciate it especially given that it's useful work that is also quite tedious to do by oneself :). I'll certainly need to refresh my memory a bit as it's been a while since I haven't touched this project, but I'll definitely try to have a look soon.

Did you hit any particular issues that you'd like to report, when building the project or extending the existing bindings?

jsoo1 commented:
Hey, thanks! It was a really great learning experience. I got from 0 to 1 with the ctypes library.

The only real problem came up about 2 weeks ago with the release of wlroots 0.13. A lot of the unstable features changed.

Armael commented:
Yeah, I noticed today that the repo doesn't build anymore with the latest git version of wlroots.
I'd like to try and look at whether wlroots could be vendored with the bindings, to make it easier to rely on a fixed version.

While I figure how to do that, could you tell me which git commit of wlroots you used when working on the bindings?

jsoo1 commented:
We were using the `0.12.0` tag.

I personally try to avoid vendoring whenever possible. I am not sure how fast 0.13 is going to be adopted in distros, though.

Armael commented:
Thanks. I was thinking of vendoring only for developping the library, not for what would be a released version of the bindings.
Because otherwise I have to uninstall/downgrade my system installation of wlroots (and thus sway) to work on the bindings, which is not very convenient.

jsoo1 commented:
Fair enough. I am using nix/guix to manage the wlroots installation so it's possible to have multiple versions installed. Totally understand if that does not help your process, though.

Armael commented:
Ah, that's a very good point. Relying on nix or guix for development does sound simpler than figuring out vendoring. Would you mind sharing which setup you use for that? (if you have a nix/guix profile file it could be part of the repo)
Does that mean that you don't use opam to setup your ocaml development environment?

jsoo1 commented:
I used opam pretty extensively but I don't think it would be entirely necessary. I could setup a nix shell or guix environment if you wish. Right now my setup is a bit of a patchwork.

Armael commented:
Sorry for the delay, I just came back from holidays. I think if you could cook up a nix or guix environment that would be super useful (even more if it somehow works with opam, as that's what I know most). I tried to do that my self (with guix), but couldn't manage to make it work.

Armael commented:
I've been suggested another approach on #sway-devel: to build wlroots and install it locally (e.g. in `~/.local`), then tweak environment variables to point to the local installation. I guess that's a form of poor-man's nix environment, but if it works it would be convenient enough for me. I'll try that tonight and report back.

Armael commented on ffi/ffi.ml:
```diff
@@ -130,6 +134,17 @@ struct
+  let wlr_output_render_software_cursors = foreign "wlr_output_render_software_cursors"
+      (* FIXME: The void pointer is a pixman_region32_t for which no bindings exist (yet).
+         This is only ok because so far, no one uses it. *)
+      (wlr_output_p @-> ptr void @-> returning void)
```
I would instead add a `let pixman_region32_p = ptr void` somewhere earlier in the file and use that here (maybe with a comment saying that it represents a pointer to a `pixman_region32_t`).

jsoo1 replied:
+1

jsoo1 replied:
Fixed!

Armael commented on ffi/ffi.ml:
```diff
@@ -104,6 +104,10 @@ struct
 
+  (* wl_output_transform *)
+
+  let wl_output_transform = Wl_output_transform.t
```
I would drop the alias here (it's not adding anything) and directly use `Wl_output_transform.t` below.

jsoo1 replied:
Fixed!

Armael commented on lib/cursor.ml:
```diff
@@ -7,33 +7,40 @@ module Types = Wlroots_ffi_f.Ffi.Types
+let move = Bindings.wlr_cursor_move
 
-let signal_motion (cursor: t) : Pointer.Event_motion.t Wl.Signal.t = {
+let signal_motion (cursor: t) : Event_pointer_motion.t Wl.Signal.t = {
```
Why move these to external modules? The naming convention I was trying to follow (and that also appears with Keyboard.Event_key for example) is that if there's a wlr_foo.h, then `wlr_event_foo_bar` gets mapped to `Foo.Event_bar`.
Did that naming scheme not work here? Otherwise I'd prefer to stick to it.

jsoo1 replied:
I think we did this because we were struggling to learn the module system.  We can definitely use a different convention.

I was kind of hoping to move events to submodules as a convention, like `Pointer.Event.motion` for instance.

Armael replied:
I think in any case each event type needs to have its own module (so it can include its compare/hash function for example), so it would have to be `Pointer.Event.Motion.t`.
I think having an `Event` submodule would be useful if we had functions that worked on all events, so that we could put them there (`Pointer.Event.foo`). However I IIRC don't think we have such functions currently, so for now I would just stick with the `Pointer.Event_foo` naming scheme.

Armael commented on lib/edges.ml:
```diff
@@ -0,0 +1,18 @@
+module Bindings = Wlroots_ffi_f.Ffi.Make (Generated_ffi)
+module Types = Wlroots_ffi_f.Ffi.Types
+
+type edges = None | Top | Bottom | Left | Right
```
should be `edge`

jsoo1 replied:
I agree, but we wanted to stick to the names from wlroots. I think it is defined as `wlr_edges` in https://github.com/swaywm/wlroots/blob/475d9701e21e5047ee9d7e56909762c1eb961fcd/include/wlr/util/edges.h#L19:
```C
enum wlr_edges {
	WLR_EDGE_NONE = 0,
	WLR_EDGE_TOP = 1 << 0,
	WLR_EDGE_BOTTOM = 1 << 1,
	WLR_EDGE_LEFT = 1 << 2,
	WLR_EDGE_RIGHT = 1 << 3,
};
```

Armael replied:
Ah, I see. I still think that in ocaml it would be more natural to have the type named `edge`. In C it makes sense because an element of `wlr_edges` is going to be used as a bit-field and thus indeed corresponds to a set of edges ; but this is not the case for an inhabitant of the ocaml sum type.

Armael commented on common/utils.ml:
```diff
@@ -31,6 +31,23 @@ let bitwise_enum desc =
+    ) zero items
+  in
+  view ~read ~write uint32_t
+
```
When should we use `bitwise_enum32` instead of `bitwise_enum`?
After a quick google search it looks like it's not terribly well specified which size enums should have; that they fit in a uint32_t is probably a good conservative assumption. However `enum` from ctypes assumes int64 values so this is probably fine as well?

In any case I would make a choice and either get rid of `bitwise_enum` or of `bitwise_enum32`.

jsoo1 replied:
Ah yeah, this was because the `edges` field of the `wlr_xdg_toplevel_resize_event` is a `uint32_t` here https://github.com/swaywm/wlroots/blob/84dea55b20b60c229a5e31ccd37b58c96cba611a/include/wlr/types/wlr_xdg_shell.h#L224:
```C
struct wlr_xdg_toplevel_resize_event {
	struct wlr_xdg_surface *surface;
	struct wlr_seat_client *seat;
	uint32_t serial;
	uint32_t edges;
};
```
Which is used in `begin_interactive` in tinywl https://github.com/swaywm/wlroots/blob/475d9701e21e5047ee9d7e56909762c1eb961fcd/tinywl/tinywl.c#L723:
```C
static void begin_interactive(struct tinywl_view *view,
		enum tinywl_cursor_mode mode, uint32_t edges) {
...
		double border_x = (view->x + geo_box.x) + ((edges & WLR_EDGE_RIGHT) ? geo_box.width : 0);
		double border_y = (view->y + geo_box.y) + ((edges & WLR_EDGE_BOTTOM) ? geo_box.height : 0);
...
}

static void xdg_toplevel_request_move(
		struct wl_listener *listener, void *data) {
	/* This event is raised when a client would like to begin an interactive
	 * move, typically because the user clicked on their client-side
	 * decorations. Note that a more sophisticated compositor should check the
	 * provied serial against a list of button press serials sent to this
	 * client, to prevent the client from requesting this whenever they want. */
	struct tinywl_view *view = wl_container_of(listener, view, request_move);
	begin_interactive(view, TINYWL_CURSOR_MOVE, 0);
}

static void xdg_toplevel_request_resize(
		struct wl_listener *listener, void *data) {
	/* This event is raised when a client would like to begin an interactive
	 * resize, typically because the user clicked on their client-side
	 * decorations. Note that a more sophisticated compositor should check the
	 * provied serial against a list of button press serials sent to this
	 * client, to prevent the client from requesting this whenever they want. */
	struct wlr_xdg_toplevel_resize_event *event = data;
	struct tinywl_view *view = wl_container_of(listener, view, request_resize);
	begin_interactive(view, TINYWL_CURSOR_RESIZE, event->edges);
}

```
I wonder if this is an upstream issue.

jsoo1 replied:
Another related question: Should `bitwise_enum` be defined in terms of `int64` instead of `uint64_t`? (due to the underspecified nature of enum sizes)

Armael replied:
Thanks for the details. Then I agree that sticking to the C api and avoiding any integer casts is best. In that case it makes sense to have both `bitwise_enum` and `bitwise_enum32`.

john-consumable replied:
Do you think it would be possible to use a functor to parameterize the type of number (i.e. uint32)?

jsoo1 replied:
Would it be possible to use a functor to parameterize the ~type of uint~ size of the enum members?

jsoo1 replied:
I just pushed a WIP commit with a functor for bitwise enums. It does change uint64 to int64 where the unsigned version was used.

Armael commented on lib/keyboard.ml:
```diff
@@ -21,9 +21,27 @@ end
+      c = keyboard |-> Types.Keyboard.events_modifiers;
+      typ = ptr Types.Keyboard.t;
+  }
+end
```
I agree with the change in naming convention proposed here (`signal_foo` -> `Events.foo`). However then it needs to be done consistently, right now it looks like there are `signal_*` functions remaining (e.g. in cursor.ml)?

jsoo1 replied:
Would you like the renaming done as part of this patch series? I think we can do that if so.

Armael commented on lib/keyboard_modifiers.ml:
```diff
@@ -0,0 +1,14 @@
+  Signed.Int64.of_int 0 !=
+    Signed.Int64.logand
+       (Unsigned.UInt32.to_int64 modifiers)
+       Types.Keyboard_modifier._WLR_MODIFIER_ALT
```
Why not have this module be a submodule of `Keyboard` instead? (`Keyboard.Modifiers`)

jsoo1 replied:
:sweat_smile: yeah, we were struggling with the module system. We'll definitely do that

jsoo1 replied:
Fixed!

Armael commented on lib/keyboard_modifiers.ml:
```diff
@@ -0,0 +1,14 @@
+type t = Types.Keyboard_modifiers.t ptr
+include Ptr
+
+let has_alt modifiers =
```
I assume this is WIP?

jsoo1 replied:
Yes! This was before we knew about `bitwise_enum` I think.

Armael commented on lib/keycodes.ml:
```diff
@@ -0,0 +1,7 @@
+module Types = Wlroots_ffi_f.Ffi.Types
+
+type t = Types.Keycodes.t
+include Ptr
```
Unless really needed I would try to avoid exposing Keycodes.t as its own type, and instead convert from/to lists of keycodes(=integers) if we can get away with that.

Armael commented on types/types.ml:
```diff
@@ -40,9 +40,32 @@ module Make (S : Cstubs_structs.TYPE) = struct
+                   | FlippedOneEighty
+                   | FlippedTwoSeventy
+
+    type t = [`wl_output_transform] Ctypes.structure
```
I suspect this can be removed?

jsoo1 replied:
I'm investigating this. I'll get back to you. What makes you think it can be removed?

Armael commented on types/types.ml:
```diff
@@ -40,9 +40,32 @@ module Make (S : Cstubs_structs.TYPE) = struct
-    (* FIXME *)
-    type t = int64
-    let t : t typ = int64_t
+    type transform = Ninety
```
this type should be named `t` for consistency

Armael commented on types/types.ml:
```diff
@@ -40,9 +40,32 @@ module Make (S : Cstubs_structs.TYPE) = struct
+                   | Flipped
+                   | FlippedNinety
+                   | FlippedOneEighty
+                   | FlippedTwoSeventy
```
To stay as close as possible to the C enum names, I think I would rather try to keep the numbers, e.g. `T90`, `T180`, `T270`, and then `Flipped_90`, etc.

Armael commented on lib/seat.ml:
```diff
@@ -49,3 +66,37 @@ let signal_request_set_cursor (seat: t) : _ Wl.Signal.t = {
+    seat
+    (Keyboard.Event_key.time_msec evt)
+    (Unsigned.UInt32.of_int (Keyboard.Event_key.keycode evt))
+    (coerce Types.Key_state.t uint32_t (Keyboard.Event_key.state evt))
```
It looks like this function is doing more work than it should? Why does it take a Keyboard.Event_key.t instead of arguments similar to the ones of the C function?

Armael commented on lib/wl.ml:
```diff
@@ -115,11 +115,6 @@ module Resource = struct
-module Output_transform = struct
-  type t = Types.Wl_output_transform.t
-  include Poly
-end
```
Why move this out of this module? The intent here is that if something is defined in the wayland libs (not wlroots), then it should be in the Wl module.

Armael commented on lib/wlroots.mli:
```diff
@@ -218,37 +273,90 @@ module Compositor : sig
 end
 
-module Xdg_shell : sig
+module rec Xdg_toplevel: sig
```
`module rec` looks scary, but after some thought I think this is a quite neat solution (especially considering the fact that the rec is only at the level of the module signatures, not the implementation).

An alternative would be to have a `Xdg` module, with:
```
module Xdg : sig
  type surface
  type toplevel
  type shell

  module Surface : sig
    type t = surface
   ...
  end

  module Toplevel = sig
    type t = toplevel 
    ...
  end

  ...

end
```

Armael replied:
haven't made my mind as to whether I prefer the solution with the `rec` or the `Xdg` module... thoughts?

Armael commented on lib/wlroots.mli:
```diff
@@ -218,37 +273,90 @@ module Compositor : sig
-  module Surface : sig
-    include Comparable0
+  module Events: sig
+    module Move : sig
```
this is not following the naming convention we've seen earlier afaict.
I would expect an `Events` module contain `signal` fields.
And for the structures describing the payload of a given event, there should be `Event_foo` modules (outside of `Events`).

Armael commented on types/types.ml:
```diff
@@ -285,10 +422,47 @@ module Make (S : Cstubs_structs.TYPE) = struct
+
+    type role = None | TopLevel | Popup
+
+    type t = [`xdg_surface_role] Ctypes.structure
```
this can be removed I think?

Armael commented on types/types.ml:
```diff
@@ -285,10 +422,47 @@ module Make (S : Cstubs_structs.TYPE) = struct
+    let top_level = constant "WLR_XDG_SURFACE_ROLE_TOPLEVEL" int64_t
+    let popup = constant "WLR_XDG_SURFACE_ROLE_POPUP" int64_t
+
+    type role = None | TopLevel | Popup
```
this should be `t`

Armael commented on lib/wlroots.mli:
```diff
@@ -218,37 +273,90 @@ module Compositor : sig
 
   val create : Wl.Display.t -> t
-  val signal_new_surface : t -> Surface.t Wl.Signal.t
+  val signal_new_surface : t -> Xdg_surface.t Wl.Signal.t
```
According to the new naming scheme, this should be `Events.new_surface`

jsoo1 commented:
I can certainly setup a nix/guix environment. It might take a week or so,
I'm quite busy right now

On Thu, Jul 8, 2021 at 3:58 AM Armaël Guéneau ***@***.***>
wrote:

> I've been suggested another approach on #sway-devel: to build wlroots and
> install it locally (e.g. in ~/.local), then tweak environment variables
> to point to the local installation. I guess that's a form of poor-man's nix
> environment, but if it works it would be convenient enough for me. I'll try
> that tonight and report back.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/swaywm/wlroots-ocaml/pull/5*issuecomment-876341610__;Iw!!IKRxdwAv5BmarQ!M_CZEjO-9gbj1Pv9FiJ4jZ1mfNt4_EjVLU9Ot3BM65kXOSlUPe1ijsX7L3VNVQ$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ACMTD2N7ASVUCJD4IUKLRXTTWWAEVANCNFSM46T7KZLQ__;!!IKRxdwAv5BmarQ!M_CZEjO-9gbj1Pv9FiJ4jZ1mfNt4_EjVLU9Ot3BM65kXOSlUPe1ijsV-QEgmMw$>
> .
>

Armael commented:
Alright, don't worry too much then, it turns out the solution of installing wlroots in ~/.local works well for me; I could compile your branch and I'm currently doing the review :-).

jsoo1 commented:
Ok thanks! There are still a few bugs with it but thank you!

> On Friday, Jul 09, 2021 at 1:23 PM, Armaël Guéneau ***@***.*** ***@***.***)> wrote:
>
>
>
>
> Alright, don't worry too much then, it turns out the solution of installing wlroots in ~/.local works well for me; I could compile your branch and I'm currently doing the review :-).
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub (https://urldefense.com/v3/__https://github.com/swaywm/wlroots-ocaml/pull/5*issuecomment-877436533__;Iw!!IKRxdwAv5BmarQ!P2bGYfsMd8BqVtRDi7swmpCjCjJ2SqSGVs58PmwvndJGPy-ZaAtL-QFcMvX5yg$), or unsubscribe (https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ACMTD2KYOKA2F2X664WP56LTW5LDDANCNFSM46T7KZLQ__;!!IKRxdwAv5BmarQ!P2bGYfsMd8BqVtRDi7swmpCjCjJ2SqSGVs58PmwvndJGPy-ZaAtL-QE2Pdo2nQ$).
>
>
>

Armael commented on lib/xdg_surface.ml:
```diff
@@ -0,0 +1,85 @@
+include Ptr
+
+let role (surface : t) : role =
+  surface |->> Types.Xdg_surface.role
```
this is `getfield Types.Xdg_surface.role` isn't it?

Armael commented on lib/xdg_surface.ml:
```diff
@@ -0,0 +1,85 @@
+  (* Worth it? *)
+  if Bindings.wlr_surface_is_xdg_surface surface
+  then
+    (* assert is called so this might blow up *)
```
I'm not sure I understand the comment there?

Armael commented on lib/xdg_surface.ml:
```diff
@@ -0,0 +1,85 @@
+let surface = getfield Types.Xdg_surface.surface
+let toplevel = getfield Types.Xdg_surface.toplevel
+
+let from_surface (surface : Surface.t) : t option =
```
Should this function be called `from_wlr_surface` for consistency?

Armael commented on lib/xdg_toplevel.ml:
```diff
@@ -0,0 +1,42 @@
+    let seat = getfield Types.Xdg_toplevel_resize_event.seat
+    let serial = getfield Types.Xdg_toplevel_resize_event.serial
+    let edges ev = coerce uint32_t Edges.t
+                     (getfield Types.Xdg_toplevel_resize_event.edges ev)
```
instead of coerce, shouldn't this be using bitwise_enum?

Armael commented on types/types.ml:
```diff
@@ -144,12 +190,47 @@ module Make (S : Cstubs_structs.TYPE) = struct
   end
 
+  module Keyboard_modifier = struct
+    type modifier =
```
I would name this `t`

Armael commented on types/types.ml:
```diff
@@ -144,12 +190,47 @@ module Make (S : Cstubs_structs.TYPE) = struct
+    let _WLR_MODIFIER_LOGO = constant "WLR_MODIFIER_LOGO" int64_t
+    let _WLR_MODIFIER_MOD5 = constant "WLR_MODIFIER_MOD5" int64_t
+
+    let modifier : modifier typ =
```
I would also name this `t`

Armael reviewed (commented):
Here's a first round of review (I haven't looked at tinywl itself yet). It's looking good, I only have a number of comments mostly regarding naming... Thanks for the hard work!

jsoo1 committed:
ffi: Add pixman_region32_t.

Note that pixman_region32_t is stubbed as a void pointer since no
bindings to pixman exist yet.

jsoo1 committed:
ffi: Remove wl_output_transform.

Just use Wl_output_transform.t, as wl_output_transform was just an alias.

jsoo1 committed:
lib: Move Keyboard_modifiers to submodule of Keyboard.

jsoo1 commented:
Thanks for the detailed review! I'll be going over it this week.

jsoo1 committed:
common: Add Bitwise functor for bitwise enums.

jsoo1 committed:
[WIP] bitwise_enums: Use Bitwise.Enum modules instead of functions
